substrate/client/rpc-servers/src/middleware/chain_head.rs

```diff
@@ -0,0 +1,237 @@
+	}
+}
+
+impl<'a, S> RpcServiceT<'a> for ChainHeadMiddleware<S>
+where
+	S: Send + Sync + RpcServiceT<'a>,
+{
+	type Future = ResponseFuture<S::Future>;
+
+	fn call(&self, req: Request<'a>) -> Self::Future {
+		const CHAIN_HEAD_FOLLOW: &str = "chainHead_unstable_follow";
+		const CHAIN_HEAD_CALL_METHODS: [&str; 8] = [
+			"chainHead_unstable_body",
+			"chainHead_unstable_header",
+			"chainHead_unstable_call",
+			"chainHead_unstable_unpin",
+			"chainHead_unstable_continue",
+			"chainHead_unstable_storage",
+			"chainHead_unstable_stopOperation",
+			"chainHead_unstable_unfollow",
+		];
+
+		let method_name = req.method_name();
+
+		// Intercept the subscription ID returned by the `chainHead_follow` method.
+		if method_name == CHAIN_HEAD_FOLLOW {
+			return ResponseFuture::Register {
+				fut: self.service.call(req.clone()),
+				connection_data: self.connection_data.clone(),
+			}
+		}
+
+		// Ensure the subscription ID of those methods corresponds to a subscription ID
+		// of this connection.
+		if CHAIN_HEAD_CALL_METHODS.contains(&method_name) {
+			let params = req.params();
+			let follow_subscription = get_subscription_id(params);
+
+			if let Some(follow_subscription) = follow_subscription {
+				if !self.connection_data.lock().subscriptions.contains(&follow_subscription) {
+					return ResponseFuture::Ready {
+						response: Some(MethodResponse::error(
+							req.id(),
+							jsonrpsee::types::error::ErrorObject::owned(
+								-32602,
+								"Invalid subscription ID",
+								None::<()>,
+							),
+						)),
+					};
+				}
+			}
+		}
+
+		ResponseFuture::Forward { fut: self.service.call(req.clone()) }
+	}
+}
+
+/// Extract the subscription ID from the provided parameters.
+///
+/// We make the assumption that all `chainHead` methods are given the
+/// subscription ID as a first parameter.
+///
+/// This method handles positional and named `camelCase` parameters.
+fn get_subscription_id<'a>(params: Params<'a>) -> Option<String> {
+	// Support positional parameters.
+	if let Ok(follow_subscription) = params.sequence().next::<String>() {
+		return Some(follow_subscription)
+	}
+
+	// Support named parameters.
+	let Ok(value) = params.parse::<serde_json::Value>() else { return None };
+
+	let serde_json::Value::Object(map) = value else { return None };
+	if let Some(serde_json::Value::String(subscription_id)) = map.get("followSubscription") {
+		return Some(subscription_id.clone())
+	}
+
+	None
+}
+
+/// Extract the result of a jsonrpc object.
+///
+/// The function extracts the `result` field from the JSON-RPC response.
+///
+/// In this example, the result is `tfMQUZekzJLorGlR`.
+/// ```ignore
+/// "{"jsonrpc":"2.0","result":"tfMQUZekzJLorGlR","id":0}"
+/// ```
+fn get_method_result(response: &MethodResponse) -> Option<String> {
+	if response.is_error() {
+		return None
+	}
+
+	let result = response.as_result();
+	let Ok(value) = serde_json::from_str(result) else { return None };
```

Comment:
yeah, this is awkward to decode the response from the final JSON string.

It would be cleaner and easier with a type that implements deserialize.

```
    #[derive(serde::Deserialize)]
    struct SubscriptionPayload {
        result: String,
    }
```

I guess the rpc v2 requires to subscription ID to be a string but jsonrpsee accepts integers as well.